### ISSUE TITLE
Microsoft.CodeAnalysis.CSharp 4.13.0 causes build error Error CS0234 : The type or namespace name .. does not exist in the namespace .. (are you missing an assembly reference?)

### ISSUE THREAD

**Issue by tomohisa** [dotnet/roslyn#77472]
**Version Used**: 
Microsoft.CodeAnalysis.CSharp 4.13.0 

**Steps to Reproduce**:

1. clone https://github.com/J-Tech-Japan/Sekiban/commit/3e81ce9ad87c07a743341ef1ffe8594cf543b47f
2.  [‎src/Sekiban.Pure.SourceGenerator/Sekiban.Pure.SourceGenerator.csproj](https://github.com/dotnet/roslyn/issues#diff-81d084401c54edef6ea6395eceb84dc056a72834645ae47728b4d25305dacb31) 4.12.0 Microsoft.CodeAnalysis.CSharp does not raise error
3. [‎src/Sekiban.Pure.SourceGenerator/Sekiban.Pure.SourceGenerator.csproj](https://github.com/dotnet/roslyn/issues#diff-81d084401c54edef6ea6395eceb84dc056a72834645ae47728b4d25305dacb31) 4.13.0 Microsoft.CodeAnalysis.CSharp DOES raise error
 
A minimal repro, with source-code provided, is ideal.  Most compiler/language issues can be distilled into a snippet that can be pasted into [sharplab](https://sharplab.io/).

<img width="1548" alt="Image" src="https://github.com/user-attachments/assets/4927642c-7d7a-4590-a610-8a9c022e0a7b" />

**Diagnostic Id**:

If this is a report about a bug in an analyzer, please include the diagnostic ID and message if possible (e.g. `"IDE0030: Use coalesce expression"`).

**Expected Behavior**:

can reference generated code.

**Actual Behavior**:

when I use 4.13.0, it can not reference Generated code.

**Comment by jjonescz** [comment]
Are there other warnings, perhaps some that would say that the source generator did not run for some reason?

**Comment by RikkiGibson** [comment]
Attaching [binlogs](https://msbuildlog.com/) of the successful and failing builds to this issue may help us diagnose this a little faster (without needing to take time to set up the repo+environment ourselves, etc.)

**Comment by tomohisa** [comment]
Thanks! I am out of the desk I'll try binlogs later.

**Comment by jaredpar** [comment]
Rather than a binlog, please include a [compiler log](https://github.com/jaredpar/complog?tab=readme-ov-file#creating-compiler-logs). That makse it trivial for us to repro these types of issues

**Comment by tomohisa** [comment]
https://www.dropbox.com/scl/fi/1wyneevur12nrvx8llcon/complogs.zip?rlkey=sfhqpvlq420s4jh3sp5x1rapd&dl=0
this is complog with succeed build with 4.12.0 and failed build with 4.13.0

**Comment by tomohisa** [comment]
let me know if you need more information

**Comment by jaredpar** [comment]
Thanks for the complog, that helped me root cause the problem. 

At build time you are using the 9.0.102 version of the compiler: 

```txt
C:\Users\jaredpar\temp\repro
⚡🔨 > complog print -c .\msbuild.4.12.0.complog
...
Compilers
        File Path: /usr/local/share/dotnet/sdk/9.0.102/Roslyn/bincore/csc.dll
        Assembly Name: csc, Version=4.12.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        Commit Hash: dfa7fc6bdea31a858a402168384192b633c811fa
```

That corresponds with the 4.12.0.0 version of the compiler. When your source generator references 4.12.0.0 then everything builds because it's also loaded into a 4.12.0.0 version of the compiler. When you change the source generator to reference 4.13.0.0 then it gets loaded into a 4.12.0.0 version of the compiler which doesn't work. The compiler will reject analyzers / generators that were compiled against newer versions of the compiler: that could be depending on APIs, types, etc ... that don't exist. 

This results in a warning at build time: 

```
  Pure.Domain net9.0 succeeded with 1 warning(s) (2.5s) → C:\Users\jaredpar\temp\code\Sekiban\internalUsages\Pure.Domain\bin\Debug\net9.0\Pure.Domain.dll
    CSC : warning CS9057: The analyzer assembly 'C:\Users\jaredpar\temp\code\Sekiban\src\Sekiban.Pure.SourceGenerator\bin\Debug\netstandard2.0\Sekiban.Pure.SourceGenerator.dll' references version '4.13.0.0' of the compiler, which is newer than the currently running version '4.12.0.0'.
```

To fix this you need to start building with a 9.0.200 SDK as that uses the 4.13.0.0 compiler. 

Note: your global.json currently specifies 8.0.0.0 which could also lead to this problem. At the least I would update this to be 9.0.100 if you stay with 4.12.0.0 references.

**Comment by tomohisa** [comment]
Thank you for taking time! I updated 9.0.201 and it builds successfully!
